spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+committed state" during `Commit`. Execution of a proposed block (via `PrepareProposal`/`ProcessProposal`) must not update the `ExecuteTxState`, but rather be kept as a separate candidate state until `FinalizeBlock`
+confirms which of the candidate states (if any) can be used to update `ExecuteTxState`.
+
+#### Mempool Connection
+
+The mempool Connection should maintain a `CheckTxState`
+to sequentially process pending transactions in the mempool that have
+not yet been committed. It should be initialized to the latest committed state
+at the end of every `Commit`.
+
+During the execution of a Consensus instance, the `CheckTxState` may be updated concurrently with the
+`ExecuteTxState`, as messages may be sent concurrently on the Consensus and Mempool connections.
+At the end of the Consensus instance, as described above, Tendermint will lock the mempool and flush
+the mempool connection before calling `Commit`, ensuring that all existing `CheckTx` calls are
+responded to and no new ones can begin.
+
+After `Commit`, while still holding the mempool lock, `CheckTx` is run again on all transactions
+that remain in the node's local mempool after filtering those included in the block.
+An additional `Type` parameter is made available to the `CheckTx` function that
+indicates whether an incoming transaction is new (`CheckTxType_New`), or a
+recheck (`CheckTxType_Recheck`).
+
+Finally, after re-checking transactions in the mempool, Tendermint will unlock
+the mempool connection. New transactions are once again able to be processed through `CheckTx`.
+
+Note that `CheckTx` is just a weak filter to keep invalid transactions out of the blockchain.
+Since the transaction cannot be guaranteed to be checked against the exact same state as it
+will be executed as part of a (potential) decided block, `CheckTx` shouldn't check *everything*
+that affects the transaction's validity, in particular those checks whose validity may depend on
+transaction ordering. `CheckTx` is weak because a Byzantine node need not care about `CheckTx`;
+it can propose a block full of invalid transactions if it wants. The mechanism ABCI++ has
+in place for dealing with such behavior is `ProcessProposal`.
+
+##### Replay Protection
+
+To prevent old transactions from being replayed, `CheckTx` must implement
+replay protection.
+
+It is possible for old transactions to be sent to the application. So
+it is important `CheckTx` implements some logic to handle them.
+
+#### Info/Query Connection
+
+The Info (or Query) Connection should maintain a `QueryState` for answering queries from the user,
+and for initialization when Tendermint first starts up (both described further
+below).
+
+`QueryState` always contain a copy of of the latest `ExecuteTxState` at the end of every `Commit`,
+after the full block has been processed and the state committed to disk.
+Otherwise it should never be modified.
+
+Tendermint Core currently uses the Query connection to filter peers upon
```

Comment:
This is an example of user queries/commands, correct?

Reply:
Yeah, for some reason, it was included in this section and I didn't dare to delete it.
I guess it's important because it's a "special" example that affects p2p, and not an external user as in the normal case.

Reply:
I'd put it in it's own section. It doesn't really make sense here.

Reply:
OK, I'll delete it, then